src/webviews/plus/composer/utils/composer.utils.ts

```diff
@@ -393,5 +574,178 @@ export function validateResultingDiff(
 	includeUnstagedChanges: boolean,
 ): boolean {
 	const { hashes } = safetyState;
+
+	if (hashes.commits) {
+		return diffHash === hashes.commits;
+	}
+
+	// Working directory mode: validate against staged/unified hash
 	return diffHash === (includeUnstagedChanges ? hashes.unified : hashes.staged);
 }
+
+/**
+ * Gets commits that are unique to a branch by finding the merge base and getting commits between branch head and merge base
+ */
+export async function getBranchCommits(
+	_container: Container,
+	repo: Repository,
+	branchName: string,
+	mergeTargetName?: string,
+): Promise<{ commits: GitCommit[]; baseCommit: { sha: string; message: string }; headCommitSha: string } | undefined> {
+	try {
+		// Get the branch
+		const branch = await repo.git.branches.getBranch(branchName);
+		if (!branch) {
+			return undefined;
+		}
+
+		// Get the merge target branch
+		let baseBranch;
+		if (mergeTargetName) {
+			baseBranch = await repo.git.branches.getBranch(mergeTargetName);
+		}
+
+		if (!baseBranch) {
+			return undefined;
+		}
+
+		// Get the merge base between the branch and its target
+		const mergeBase = await repo.git.refs.getMergeBase(branch.ref, baseBranch.ref);
+		if (!mergeBase) {
+			return undefined;
+		}
+		// Get the base commit from the merge base
+		const baseCommit = await repo.git.commits.getCommit(mergeBase);
+		if (!baseCommit) {
+			return undefined;
+		}
+
+		// Get commits between merge base and branch head (excluding merge base)
+		const log = await repo.git.commits.getLog(`${baseBranch.ref}..${branch.ref}`, { limit: 0 });
+		if (!log?.commits?.size) {
+			return undefined;
+		}
+
+		// Convert Map to Array and keep in reverse chronological order (newest first, then reverse to oldest first for processing)
+		const commits = Array.from(log.commits.values()).reverse();
+		const headCommit = commits[commits.length - 1];
+
+		return {
+			commits: commits,
+			baseCommit: {
+				sha: baseCommit.sha,
+				message: baseCommit.message ?? '',
+			},
+			headCommitSha: headCommit?.sha ?? branch.sha,
+		};
+	} catch {
+		return undefined;
+	}
+}
+
+export function parseCoAuthorsFromGitCommit(commit: GitCommit): GitCommitIdentityShape[] {
+	const coAuthors: GitCommitIdentityShape[] = [];
+	if (!commit.message) return coAuthors;
+
+	const coAuthorRegex = /^Co-authored-by:\s*(.+?)(?:\s*<(.+?)>)?\s*$/gm;
+	let match;
+	while ((match = coAuthorRegex.exec(commit.message)) !== null) {
+		const [, name, email] = match;
+		if (name) {
+			coAuthors.push({ name: name.trim(), email: email?.trim(), date: commit.date });
+		}
+	}
+
+	return coAuthors;
+}
```

Comment:
This will be great to add in other commit authorship/avatar content throughout GitLens.